nvtabular/io/avro.py

```diff
@@ -0,0 +1,138 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import cudf
+import uavro as ua
+from dask.base import tokenize
+from dask.dataframe.core import new_dd_object
+
+from .dataset_engine import DatasetEngine
+
+
+class AvroDatasetEngine(DatasetEngine):
+    """AvroDatasetEngine
+
+    Uses `uavro` to decompose dataset into groups of avro blocks.
+    Uses `cudf` to create new partitions.
+    """
+
+    def __init__(self, paths, part_size, storage_options=None, **kwargs):
+        super().__init__(paths, part_size, storage_options)
+        if kwargs != {}:
+            raise ValueError("Unexpected AvroDatasetEngine argument(s).")
+        self.blocksize = part_size
+
+        # Avro reader needs a list of files
+        # (Assume flat directory structure if this is a dir)
+        if len(self.paths) == 1 and self.fs.isdir(self.paths[0]):
+            self.paths = self.fs.glob(self.fs.sep.join([self.paths[0], "*"]))
+
+    def to_ddf(self, columns=None):
+
+        # Get list of pieces for each output
+        pieces, meta = self.process_metadata(columns=columns)
+
+        # Construct collection
+        token = tokenize(self.fs, self.paths, self.part_size, columns)
+        read_avro_name = "read-avro-partition-" + token
+        dsk = {
+            (read_avro_name, i): (AvroDatasetEngine.read_partition, self.fs, piece, columns)
+            for i, piece in enumerate(pieces)
+        }
+        return new_dd_object(dsk, read_avro_name, meta.iloc[:0], [None] * (len(pieces) + 1))
+
+    def process_metadata(self, columns=None):
+
+        with open(self.paths[0], "rb") as fo:
+            header = ua.core.read_header(fo)
+
+            # Use first block for metadata
+            num_rows = header["blocks"][0]["nrows"]
+            file_byte_count = header["blocks"][0]["size"]
+            meta = cudf.io.read_avro(self.paths[0], skiprows=0, num_rows=num_rows)
+
+            # Convert the desired in-memory GPU size to the expected
+            # on-disk storage size (blocksize)
+            df_byte_count = meta.memory_usage(deep=True).sum()
+            self.blocksize = int(float(file_byte_count) / df_byte_count * self.part_size)
+
+        # Break apart files at the "Avro block" granularity
+        pieces = []
+        for path in self.paths:
+            file_size = self.fs.du(path)
+            if file_size > self.blocksize:
+                with open(path, "rb") as fo:
+                    header = ua.core.read_header(fo)
+                    ua.core.scan_blocks(fo, header, file_size)
+                    blocks = header["blocks"]
+
+                    (file_row_offset, part_row_count) = (0, 0)
+                    (file_block_offset, part_block_count) = (0, 0)
+                    (file_byte_offset, part_byte_count) = (blocks[0]["offset"], 0)
```

Comment:
nitpick: the parenthesis around each tuple isn't necessary:

```suggestion
                    file_row_offset, part_row_count = 0, 0
                    file_block_offset, part_block_count = 0, 0
                    file_byte_offset, part_byte_count = blocks[0]["offset"], 0
```